File: python/evaluation.py
from pathlib import Path
import numpy as np
import skimage as ski

from python.square_detection import should_remove_smaller, filter_contained_triplets
from square_detection import square_filter, get_triplets, get_qr_corners, get_center
from main import load_image, preprocess, denoising, labels,save_debug

def parse_ground_truth(txt_path: Path) -> list[list[tuple]]:
    qr_codes = []
    in_sets = False
    with open(txt_path, "r") as f:
        for line in f:
            line = line.strip()
            if not line or line.startswith("#"):
                continue
            if line == "SETS":
                in_sets = True
                continue
            if in_sets:
                values = list(map(float, line.split()))
                if len(values) == 8:
                    pts = [(values[i + 1], values[i]) for i in range(0, 8, 2)]
                    qr_codes.append(pts)
    return qr_codes


def corners_to_bbox(corners: list[tuple]) -> tuple:
    rows = [p[0] for p in corners]
    cols = [p[1] for p in corners]
    return (min(rows), min(cols), max(rows), max(cols))


def detect_qr(image_path: str) -> list[tuple]:
    im = load_image(image_path)
    if im.ndim == 3 and im.shape[2] == 4:
        im = im[:, :, :3]

    equalized, gray, binary = preprocess(im)
    denoise = denoising(binary)
    lab = labels(denoise)
    regions = ski.measure.regionprops(lab)

    elements = []
    for region in regions:
        res, coords = square_filter(denoise, binary, region)
        if res:
            elements.append({"bbox": region.bbox, "corners": coords, "area": region.area})

    triplets = get_triplets(elements,im.shape)
    triplets = filter_contained_triplets(triplets)
    detections = []
    for triplet in triplets:
        qr_corners = get_qr_corners(triplet)
        rows = [p[0] for p in qr_corners]
        cols = [p[1] for p in qr_corners]
        bbox = (min(rows), min(cols), max(rows), max(cols))
        detections.append(bbox)
    save_debug(Path(image_path).stem, im,equalized,gray,binary,denoise,lab,regions)

    return detections


def compute_iou(box1: tuple, box2: tuple) -> float:
    aminr, aminc, amaxr, amaxc = box1
    bminr, bminc, bmaxr, bmaxc = box2
    xA = max(aminc, bminc)
    yA = max(aminr, bminr)
    xB = min(amaxc, bmaxc)
    yB = min(amaxr, bmaxr)
    inter = max(0, xB - xA) * max(0, yB - yA)
    areaA = (amaxc - aminc) * (amaxr - aminr)
    areaB = (bmaxc - bminc) * (bmaxr - bminr)
    union = areaA + areaB - inter
    if union == 0:
        return 0.0
    return inter / union


def evaluate(detections: list, ground_truth: list, iou_threshold: float = 0.5) -> tuple[int, int, int]:
    matched_gt = set()
    tp = 0
    fp = 0
    for det in detections:
        best_iou = 0.0
        best_gt = -1
        for i, gt in enumerate(ground_truth):
            if i in matched_gt:
                continue
            iou = compute_iou(det, gt)
            if iou > best_iou:
                best_iou = iou
                best_gt = i
        if best_iou >= iou_threshold:
            tp += 1
            matched_gt.add(best_gt)
        else:
            fp += 1
    fn = len(ground_truth) - len(matched_gt)
    return tp, fp, fn


def precision(tp: int, fp: int) -> float:
    return tp / (tp + fp) if (tp + fp) > 0 else 0.0


def recall(tp: int, fn: int) -> float:
    return tp / (tp + fn) if (tp + fn) > 0 else 0.0


def f1_score(tp: int, fp: int, fn: int) -> float:
    p = precision(tp, fp)
    r = recall(tp, fn)
    return 2 * p * r / (p + r) if (p + r) > 0 else 0.0


def mean_iou(detections: list, ground_truth: list) -> float:
    scores = []
    used = set()
    for det in detections:
        best = 0.0
        best_gt = -1
        for i, gt in enumerate(ground_truth):
            if i in used:
                continue
            iou = compute_iou(det, gt)
            if iou > best:
                best = iou
                best_gt = i
        if best_gt != -1:
            used.add(best_gt)
            scores.append(best)
    return float(np.mean(scores)) if scores else 0.0


def find_gt_file(image_path: Path) -> Path | None:
    candidates = [
        image_path.with_suffix(".txt"),
        image_path.parent / (image_path.stem + ".txt"),
    ]
    for c in candidates:
        if c.exists():
            return c
    return None


def evaluate_folder(folder_path: str, iou_threshold: float = 0.5, verbose: bool = True):
    folder = Path(folder_path)
    image_extensions = {".jpg", ".jpeg", ".png", ".bmp", ".tif", ".tiff"}

    total_tp, total_fp, total_fn = 0, 0, 0
    all_ious = []
    processed = 0
    skipped = 0

    for image_file in sorted(folder.glob("*")):
        if image_file.suffix.lower() not in image_extensions:
            continue

        gt_file = find_gt_file(image_file)
        if gt_file is None:
            if verbose:
                print(f"  [SKIP] Pas de vérité terrain pour {image_file.name}")
            skipped += 1
            continue

        gt_qr_list = parse_ground_truth(gt_file)
        gt_bboxes = [corners_to_bbox(qr) for qr in gt_qr_list]

        try:
            detections = detect_qr(str(image_file))
        except Exception as e:
            import traceback
            print(f"  [ERREUR] {image_file.name} : {e}")
            traceback.print_exc()
            skipped += 1
            continue

        tp, fp, fn = evaluate(detections, gt_bboxes, iou_threshold)
        miou = mean_iou(detections, gt_bboxes)

        total_tp += tp
        total_fp += fp
        total_fn += fn
        all_ious.append(miou)
        processed += 1

        if verbose:
            p = precision(tp, fp)
            r = recall(tp, fn)
            print(
                f"  {image_file.name:40s} | "
                f"det={len(detections)} gt={len(gt_bboxes)} | "
                f"TP={tp} FP={fp} FN={fn} | "
                f"P={p:.2f} R={r:.2f} IoU={miou:.2f}"
            )

    global_p = precision(total_tp, total_fp)
    global_r = recall(total_tp, total_fn)
    global_f1 = f1_score(total_tp, total_fp, total_fn)
    global_iou = float(np.mean(all_ious)) if all_ious else 0.0

    print(f"\n{'='*60}")
    print(f"Dossier : {folder}")
    print(f"Images traitées : {processed}  |  Ignorées : {skipped}")
    print(f"{'='*60}")
    print(f"Précision globale  : {global_p:.4f}")
    print(f"Rappel global      : {global_r:.4f}")
    print(f"F1 global          : {global_f1:.4f}")
    print(f"Mean IoU global    : {global_iou:.4f}")
    print(f"{'='*60}")

    return {
        "precision": global_p,
        "recall": global_r,
        "f1": global_f1,
        "mean_iou": global_iou,
    }


if __name__ == "__main__":
    DATASET_FOLDER = "../data/Dataset/detection/monitor"
    evaluate_folder(DATASET_FOLDER, iou_threshold=0.5, verbose=True)

File: c++_implementation/src/evaluation/evaluation.cpp
#include "evaluation/evaluation.hh"

#include <algorithm>
#include <cmath>
#include <fstream>
#include <iostream>
#include <numeric>
#include <sstream>
#include <stdexcept>

#include "debug/debug_visualization.hh"
#include "detection/preprocess.hh"
#include "util/image_io.hh"

#define DEBUG false

namespace
{
    std::string strip(const std::string& s)
    {
        size_t start = s.find_first_not_of(" \t\r\n");
        if (start == std::string::npos)
            return "";
        size_t end = s.find_last_not_of(" \t\r\n");
        return s.substr(start, end - start + 1);
    }

    const std::vector<std::string> kImageExtensions = {
        ".jpg", ".jpeg", ".png", ".bmp", ".tif", ".tiff"
    };

    bool is_image_file(const std::filesystem::path& p)
    {
        std::string ext = p.extension().string();
        std::transform(ext.begin(), ext.end(), ext.begin(), ::tolower);
        return std::find(kImageExtensions.begin(), kImageExtensions.end(), ext)
            != kImageExtensions.end();
    }
} // namespace

namespace qr_code
{
    std::vector<std::vector<Point>>
    parse_ground_truth(const std::filesystem::path& txt_path)
    {
        std::vector<std::vector<Point>> qr_codes;
        std::vector<Point> current_pts;
        bool in_sets = false;

        std::ifstream file(txt_path);
        if (!file)
            throw std::runtime_error("Impossible d'ouvrir "
                                     + txt_path.string());

        std::string raw_line;
        while (std::getline(file, raw_line))
        {
            std::string line = strip(raw_line);
            if (line.empty() || line[0] == '#')
                continue;
            if (line == "SETS")
            {
                in_sets = true;
                continue;
            }

            std::istringstream iss(line);
            std::vector<double> values;
            double v;
            while (iss >> v)
                values.push_back(v);

            if (in_sets)
            {
                if (values.size() != 8)
                    throw std::runtime_error("Ligne invalide : " + line);
                std::vector<Point> pts;
                for (int i = 0; i < 8; i += 2)
                    pts.push_back({ values[i], values[i + 1] });
                qr_codes.push_back(std::move(pts));
            }
            else
            {
                if (values.size() != 2)
                    throw std::runtime_error("Ligne invalide : " + line);
                current_pts.push_back({ values[0], values[1] });
            }
        }

        if (!current_pts.empty())
        {
            if (current_pts.size() != 4)
                throw std::runtime_error("Le fichier contient "
                                         + std::to_string(current_pts.size())
                                         + " points au lieu de 4.");
            qr_codes.push_back(current_pts);
        }

        return qr_codes;
    }

    BBox corners_to_bbox(const std::vector<Point>& corners)
    {
        double min_row = corners[0].row, max_row = corners[0].row;
        double min_col = corners[0].col, max_col = corners[0].col;
        for (const Point& p : corners)
        {
            min_row = std::min(min_row, p.row);
            max_row = std::max(max_row, p.row);
            min_col = std::min(min_col, p.col);
            max_col = std::max(max_col, p.col);
        }
        return { min_row, min_col, max_row, max_col };
    }

    std::vector<BBox> detect_qr_old(const std::string& image_path)
    {
        image::rgb24_image* image = image::load_image(image_path.c_str());
        if (!image)
            throw std::runtime_error("Impossible de charger " + image_path);

        std::shared_ptr<Preprocess> pre = preprocess(*image);

        const image::gray8_image& binary = pre->image_binary;
        const image::gray8_image& denoise = pre->image_denoise;

        LabelImage lab = labels(denoise);
        std::vector<Region> regions = regionprops(lab);

        std::vector<Element> elements;
        for (const Region& region : regions)
        {
            std::vector<Point> corners;
            bool ok = square_filter(denoise, binary, region, corners);
            if (ok)
                elements.push_back({ region.bbox, corners, region.area });
        }

        std::vector<Triplet> triplets =
            get_triplets(elements, image->sx, image->sy);

        std::vector<BBox> detections;
        for (const Triplet& triplet : triplets)
        {
            std::vector<Point> qr_corners = get_qr_corners(triplet);
            if (qr_corners.empty())
                continue;

            draw_qr(*image, qr_corners);

            detections.push_back(corners_to_bbox(qr_corners));
        }

        std::filesystem::path input(image_path);
        std::filesystem::create_directories("results");

        std::filesystem::path output =
            std::filesystem::path("results") / input.filename();

        output.replace_extension(".tga");

        save_image(*image, output.string().c_str());

        delete image;
        return detections;
    }

    std::vector<BBox> detect_qr_debug(const std::string& image_path)
    {
        image::rgb24_image* image = image::load_image(image_path.c_str());

        if (!image)
            throw std::runtime_error("Impossible de charger " + image_path);

        std::shared_ptr<Preprocess> pre = preprocess(*image);
        std::string debug_folder =
            "debug/" + std::filesystem::path(image_path).stem().string();
        qr_code::save_preprocess_debug(debug_folder, *pre);
        const image::gray8_image& binary = pre->image_binary;
        const image::gray8_image& denoise = pre->image_denoise;

        LabelImage lab = labels(denoise);
        save_labels_debug(debug_folder + "/06_labels.png", lab);

        std::vector<Region> regions = regionprops(lab);
        save_regions_debug(debug_folder + "/07_regions.png", lab, regions);

        std::vector<Element> elements;
        for (const Region& region : regions)
        {
            std::vector<Point> corners;
            bool ok = square_filter(denoise, binary, region, corners);
            if (ok)
            {
                elements.push_back({ region.bbox, corners, region.area });
            }
        }
        save_elements_debug(debug_folder + "/08_elements.png", *image,
                            elements);

        std::vector<Triplet> triplets =
            get_triplets(elements, image->sx, image->sy);
        save_triplets_debug(debug_folder + "/09_triplets.png", *image,
                            triplets);

        std::vector<BBox> detections;
        std::vector<std::vector<Point>> debug_corners;
        for (const Triplet& triplet : triplets)
        {
            std::vector<Point> qr_corners = get_qr_corners(triplet);
            if (qr_corners.empty())
                continue;
            detections.push_back(corners_to_bbox(qr_corners));
            debug_corners.push_back(qr_corners);
        }

        save_qr_debug(debug_folder + "/10_qr.png", *image, debug_corners);
        delete image;
        return detections;
    }

    std::vector<BBox> detect_qr(const std::string& image_path)
    {
        if (DEBUG)
        {
            return detect_qr_debug(image_path);
        }
        else
        {
            return detect_qr_old(image_path);
        }
    }

    double compute_iou(const BBox& box1, const BBox& box2)
    {
        auto [aminr, aminc, amaxr, amaxc] = box1;
        auto [bminr, bminc, bmaxr, bmaxc] = box2;

        double xA = std::max(aminc, bminc);
        double yA = std::max(aminr, bminr);
        double xB = std::min(amaxc, bmaxc);
        double yB = std::min(amaxr, bmaxr);

        double inter = std::max(0.0, xB - xA) * std::max(0.0, yB - yA);
        double areaA = (amaxc - aminc) * (amaxr - aminr);
        double areaB = (bmaxc - bminc) * (bmaxr - bminr);
        double uni = areaA + areaB - inter;

        if (uni == 0.0)
            return 0.0;
        return inter / uni;
    }

    std::tuple<int, int, int> evaluate(const std::vector<BBox>& detections,
                                       const std::vector<BBox>& ground_truth,
                                       double iou_threshold)
    {
        std::vector<bool> matched_gt(ground_truth.size(), false);
        int tp = 0;
        int fp = 0;

        for (const BBox& det : detections)
        {
            double best_iou = 0.0;
            int best_gt = -1;
            for (size_t i = 0; i < ground_truth.size(); i++)
            {
                if (matched_gt[i])
                    continue;
                double iou = compute_iou(det, ground_truth[i]);
                if (iou > best_iou)
                {
                    best_iou = iou;
                    best_gt = static_cast<int>(i);
                }
            }
            if (best_iou >= iou_threshold)
            {
                tp++;
                matched_gt[best_gt] = true;
            }
            else
            {
                fp++;
            }
        }

        int matched_count = static_cast<int>(
            std::count(matched_gt.begin(), matched_gt.end(), true));
        int fn = static_cast<int>(ground_truth.size()) - matched_count;

        return { tp, fp, fn };
    }

    double precision(int tp, int fp)
    {
        return (tp + fp) > 0 ? static_cast<double>(tp) / (tp + fp) : 0.0;
    }

    double recall(int tp, int fn)
    {
        return (tp + fn) > 0 ? static_cast<double>(tp) / (tp + fn) : 0.0;
    }

    double f1_score(int tp, int fp, int fn)
    {
        double p = precision(tp, fp);
        double r = recall(tp, fn);
        return (p + r) > 0.0 ? 2 * p * r / (p + r) : 0.0;
    }

    double mean_iou(const std::vector<BBox>& detections,
                    const std::vector<BBox>& ground_truth)
    {
        std::vector<bool> used(ground_truth.size(), false);
        std::vector<double> scores;

        for (const BBox& det : detections)
        {
            double best = 0.0;
            int best_gt = -1;
            for (size_t i = 0; i < ground_truth.size(); i++)
            {
                if (used[i])
                    continue;
                double iou = compute_iou(det, ground_truth[i]);
                if (iou > best)
                {
                    best = iou;
                    best_gt = static_cast<int>(i);
                }
            }
            if (best_gt != -1)
            {
                used[best_gt] = true;
                scores.push_back(best);
            }
        }

        if (scores.empty())
            return 0.0;
        return std::accumulate(scores.begin(), scores.end(), 0.0)
            / static_cast<double>(scores.size());
    }

    std::optional<std::filesystem::path>
    find_gt_file(const std::filesystem::path& image_path)
    {
        std::vector<std::filesystem::path> candidates = {
            std::filesystem::path(image_path).replace_extension(".txt"),
            image_path.parent_path() / (image_path.stem().string() + ".txt")
        };
        for (const std::filesystem::path& c : candidates)
        {
            if (std::filesystem::exists(c))
                return c;
        }
        return std::nullopt;
    }

    EvalResult evaluate_folder(const std::string& folder_path,
                               double iou_threshold, bool verbose)
    {
        std::filesystem::path folder(folder_path);

        int total_tp = 0, total_fp = 0, total_fn = 0;
        std::vector<double> all_ious;
        int processed = 0;
        int skipped = 0;

        std::vector<std::filesystem::path> entries;
        for (const auto& entry : std::filesystem::directory_iterator(folder))
            entries.push_back(entry.path());
        std::sort(entries.begin(), entries.end());

        for (const std::filesystem::path& image_file : entries)
        {
            if (!is_image_file(image_file))
                continue;

            std::optional<std::filesystem::path> gt_file =
                find_gt_file(image_file);
            if (!gt_file)
            {
                if (verbose)
                    std::cout << "  [SKIP] Pas de verite terrain pour "
                              << image_file.filename().string() << "\n";
                skipped++;
                continue;
            }

            std::vector<std::vector<Point>> gt_qr_list =
                parse_ground_truth(*gt_file);
            std::vector<BBox> gt_bboxes;
            for (const std::vector<Point>& qr : gt_qr_list)
                gt_bboxes.push_back(corners_to_bbox(qr));

            std::vector<BBox> detections;
            try
            {
                detections = detect_qr(image_file.string());
            }
            catch (const std::exception& e)
            {
                std::cout << "  [ERREUR] " << image_file.filename().string()
                          << " : " << e.what() << "\n";
                skipped++;
                continue;
            }

            auto [tp, fp, fn] = evaluate(detections, gt_bboxes, iou_threshold);
            double miou = mean_iou(detections, gt_bboxes);

            total_tp += tp;
            total_fp += fp;
            total_fn += fn;
            all_ious.push_back(miou);
            processed++;

            if (verbose)
            {
                double p = precision(tp, fp);
                double r = recall(tp, fn);
                std::cout << "  " << image_file.filename().string()
                          << " | det=" << detections.size()
                          << " gt=" << gt_bboxes.size() << " | TP=" << tp
                          << " FP=" << fp << " FN=" << fn << " | P=" << p
                          << " R=" << r << " IoU=" << miou << "\n";
            }
        }

        double global_p = precision(total_tp, total_fp);
        double global_r = recall(total_tp, total_fn);
        double global_f1 = f1_score(total_tp, total_fp, total_fn);
        double global_iou = all_ious.empty()
            ? 0.0
            : std::accumulate(all_ious.begin(), all_ious.end(), 0.0)
                / static_cast<double>(all_ious.size());

        std::cout << "\n" << std::string(60, '=') << "\n";
        std::cout << "Dossier : " << folder.string() << "\n";
        std::cout << "Images traitees : " << processed
                  << "  |  Ignorees : " << skipped << "\n";
        std::cout << std::string(60, '=') << "\n";
        std::cout << "Precision globale  : " << global_p << "\n";
        std::cout << "Rappel global      : " << global_r << "\n";
        std::cout << "F1 global          : " << global_f1 << "\n";
        std::cout << "Mean IoU global    : " << global_iou << "\n";
        std::cout << std::string(60, '=') << "\n";

        return { global_p, global_r, global_f1, global_iou };
    }

} // namespace qr_code